gloo/transport/tcp/tls/pair.cc

```diff
@@ -0,0 +1,386 @@
+        return false;
+      }
+      total_rv += rv;
+      op.nwritten += rv;
+      break;
+    }
+
+  }
+  GLOO_ENFORCE_EQ(total_rv, nbytes);
+  GLOO_ENFORCE_EQ(op.nwritten, op.preamble.nbytes);
+
+  // Write completed
+  switch (opcode) {
+    case Op::SEND_BUFFER:
+      op.buf->handleSendCompletion();
+      break;
+    case Op::SEND_UNBOUND_BUFFER:
+      buf->handleSendCompletion(rank_);
+      break;
+    case Op::NOTIFY_SEND_READY:
+      break;
+    case Op::NOTIFY_RECV_READY:
+      break;
+  }
+
+  return true;
+}
+
+bool Pair::read() {
+  NonOwningPtr<UnboundBuffer> buf;
+  auto start = std::chrono::steady_clock::now();
+
+  for (;;) {
+    struct iovec iov = {
+        .iov_base = nullptr,
+        .iov_len = 0,
+    };
+    const auto nbytes = prepareRead(rx_, buf, iov);
+    if (nbytes < 0) {
+      return false;
+    }
+
+    // Break from loop if the op is complete.
+    // Note that this means that the buffer pointer has been
+    // set, per the call to prepareRead.
+    if (nbytes == 0) {
+      break;
+    }
+
+    // If busy-poll has been requested AND sync mode has been enabled for pair
+    // we'll keep spinning calling recv() on socket by supplying MSG_DONTWAIT
+    // flag. This is more efficient in terms of latency than allowing the kernel
+    // to de-schedule this thread waiting for IO event to happen. The tradeoff
+    // is stealing the CPU core just for busy polling.
+    ssize_t rv = 0;
+    for (;;) {
+      rv = SSL_read(ssl_, iov.iov_base, iov.iov_len);
+      if (rv <= 0) {
+        int err = SSL_get_error(ssl_, rv);
+
+        GLOO_ENFORCE(err != SSL_ERROR_NONE);
+        GLOO_ENFORCE(err != SSL_ERROR_WANT_WRITE);
+
+        if (err == SSL_ERROR_WANT_READ) {
+          return false;
+        }
+
+        if (err == SSL_ERROR_ZERO_RETURN) {
+          if (errno == ECONNRESET) {
+            if (!sync_) {
+              return false;
+            }
+          }
+        }
+
+        if (err == SSL_ERROR_SYSCALL) {
+          if (errno == EPIPE) {
+            if (!sync_) {
+              return false;
+            }
+          }
+          if (errno == ECONNRESET) {
+            if (!sync_) {
+              return false; // Exit if Connection reset by peer on read
+            }
+          }
+        }
+
+        // Unexpected error
+        signalException(
+            GLOO_ERROR_MSG(
+                "Read error ", peer_.str(), ": ", err, ", errno = ", strerror(errno)));
+        return false;
+      }
+      break;
+    }
+
+    // Transition to CLOSED on EOF
+    if (rv == 0) {
+      signalException(
+          GLOO_ERROR_MSG("Connection closed by peer ", peer_.str()));
+      return false;
+    }
+
+    rx_.nread += rv;
+  }
+
+  // Read completed
+  const auto opcode = rx_.getOpcode();
+  switch (opcode) {
+    case Op::SEND_BUFFER:
+      // Done sending data to pinned buffer; trigger completion.
+      rx_.buf->handleRecvCompletion();
```

Comment:
any reason TLS requires `handleRecvCompletion` to be a public API and call it here while other transports do not?

Reply:
The existing code based on friend classes and this prevents access from sub-classes